public/main.ts

```diff
@@ -2,6 +2,135 @@ let timer = 256;
+// TODO: Define different growth rates for crops?
+crops.forEach(crop => cropGrowthRate.set(crop, baseCropGrowthRate));
+
+// Price per 1 unit of crop
+// TODO: Randomly move the price up and down
+const cropMarket = new Map(
+    [
+        [Crop.Wheat, 3],
+        [Crop.Barley, 2],
+        [Crop.Rye, 1],
+    ]
+);
+
+// Food eaten by one person
+const consumptionRate = 0.5;
+
+// A farm
+class Farm {
+    totalFarmers = 0;
+    stockpile: number;
+
+    constructor(initalSeed: number){
+        this.stockpile = initalSeed;
+        console.assert(this.stockpile > 0);
+    }
+
+    // TODO: What if stockpile hits 0 and all food is eaten by the farmers
+    harvest(crop: Crop){
+        const amountGrown = Math.min(this.totalFarmers, this.stockpile);
+        // Non-null assertion as a Map may not have every entry ready
+        this.stockpile += amountGrown * cropGrowthRate.get(crop)!;
+        this.stockpile -= this.totalFarmers * consumptionRate;
+        console.assert(this.stockpile > 0);
+    }
+}
+
+// Loot is an interface b/c kingdom (currently) can become a Loot object
+interface Loot {
+    gold: number;
+    farms: Map<Crop, Farm>;
+}
+
+const baseDPSPerSoldier = 1;
+
+class Army {
+    totalSoldiers = 0;
+    dpsPerSoldier = baseDPSPerSoldier;
+
+    attack(kingdom: Kingdom): Loot {
+        //TODO: How to attack a kingdom
+        console.assert(false, "Attacking a Kingdom is not implemented");
+        return kingdom;
+    }
+}
+
+// Returns a random integer inclusive on both ends
+function getRandInt(min:number, max:number):number {
+    return Math.floor(Math.random() * (max - min + 1) ) + min;
+}
+
+
+class Kingdom {
+    name: string;
+    // TODO: Better define health
+    health = 100;
+    gold = 0;
+    idlePopulation: number;
+    army = new Army();
+    farms: Map<Crop, Farm> = new Map();
+
+    constructor(name: string, initialPopulation: number){
+        this.idlePopulation = initialPopulation;
+        this.name = name;
+    }
+
+    orderHarvest() {
+        this.farms.forEach(
+            (farm, crop) => farm.harvest(crop)
+        );
+    }
+
+    buyCrop(crop: Crop, amt: number) {
+        const cost = amt * cropMarket.get(crop)!;
+        if(cost > this.gold) return;
+
+        const farm = this.farms.get(crop)!;
+        this.gold -= cost;
+        farm.stockpile += amt;
+
+        // TODO: Increase price after purchase
+    }
+
+    sellCrop(crop: Crop, amt: number){
+        const farm = this.farms.get(crop)!;
+        if(farm.stockpile < amt) return;
+
+        const profit = amt * cropMarket.get(crop)!;
+        farm.stockpile -= amt;
+        this.gold += profit;
+
+        // TODO: Decrease price after sale
+    }
+}
+
+const minStartingRye = 50;
+const maxStartingRye = 100;
+
+const minStartingPop = 1;
+const maxStartingPop = 10;
+
+const kingdoms = [
+    // TODO: Allow player to add own name
+    new Kingdom("player", getRandInt(minStartingPop, maxStartingPop)), // player
```

Comment:
I am thinking along the lines of some function wrapping [window.prompt](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt).

Reply:
Yea I like that solution we could do that.